lib/web_ui/lib/src/engine/canvaskit/render_canvas.dart

```diff
@@ -0,0 +1,165 @@
+/// on the maximum amount of WebGL contexts which can be live at once. Using
+/// a single OffscreenCanvas and multiple RenderCanvases allows us to only
+/// create a single WebGL context.
+class RenderCanvas {
+  RenderCanvas();
+
+  /// The root HTML element for this canvas.
+  ///
+  /// This element contains the canvas used to draw the UI. Unlike the canvas,
+  /// this element is permanent. It is never replaced or deleted, until this
+  /// canvas is disposed of via [dispose].
+  ///
+  /// Conversely, the canvas that lives inside this element can be swapped, for
+  /// example, when the screen size changes, or when the WebGL context is lost
+  /// due to the browser tab becoming dormant.
+  final DomElement htmlElement = createDomElement('flt-canvas-container');
+
+  /// The underlying `<canvas>` element used to display the pixels.
+  DomCanvasElement? canvasElement;
+  int _pixelWidth = -1;
+  int _pixelHeight = -1;
+
+  DomCanvasRenderingContextBitmapRenderer? renderContext;
+
+  ui.Size? _currentCanvasPhysicalSize;
+  ui.Size? _currentRenderSize;
+  double _currentDevicePixelRatio = -1;
+
+  bool _addedToScene = false;
+  void addToScene() {
+    if (!_addedToScene) {
+      CanvasKitRenderer.instance.sceneHost!.prepend(htmlElement);
+    }
+    _addedToScene = true;
+  }
+
+  /// Sets the CSS size of the canvas so that canvas pixels are 1:1 with device
+  /// pixels.
+  ///
+  /// The logical size of the canvas is not based on the size of the window
+  /// but on the size of the canvas, which, due to `ceil()` above, may not be
+  /// the same as the window. We do not round/floor/ceil the logical size as
+  /// CSS pixels can contain more than one physical pixel and therefore to
+  /// match the size of the window precisely we use the most precise floating
+  /// point value we can get.
+  void _updateLogicalHtmlCanvasSize() {
+    final double logicalWidth = _pixelWidth / window.devicePixelRatio;
+    final double logicalHeight = _pixelHeight / window.devicePixelRatio;
+    final DomCSSStyleDeclaration style = canvasElement!.style;
+    style.width = '${logicalWidth}px';
+    style.height = '${logicalHeight}px';
+  }
+
+  /// This function is expensive.
+  ///
+  /// It's better to reuse canvas if possible.
+  void _createNewCanvas(ui.Size physicalSize) {
+    // Clear the container, if it's not empty. We're going to create a new <canvas>.
+    if (canvasElement != null) {
+      canvasElement!.remove();
+    }
+
+    // If `physicalSize` is not precise, use a slightly bigger canvas. This way
+    // we ensure that the rendred picture covers the entire browser window.
+    _pixelWidth = physicalSize.width.ceil();
+    _pixelHeight = physicalSize.height.ceil();
+    final DomCanvasElement htmlCanvas = createDomCanvasElement(
+      width: _pixelWidth,
+      height: _pixelHeight,
+    );
+    canvasElement = htmlCanvas;
+    renderContext = htmlCanvas.contextBitmapRenderer;
```

Comment:
Looking at our implementation of `get contextBitmapRenderer` we simply call `getContext('bitmaprenderer')`. According to [the spec](https://html.spec.whatwg.org/multipage/canvas.html#the-imagebitmaprenderingcontext-interface) the default value of `alpha` is `false`, which means "the canvas is forced to always be opaque". But we do want canvases to be transparent.

Reply:
Or maybe I'm wrong. It says `false` in the `getContext` signature:

```
context = canvas.getContext('bitmaprenderer' [, { [ alpha: false ] } ])
```

But the dictionary says it's `true`:

```
dictionary ImageBitmapRenderingContextSettings {
  boolean alpha = true;
};
```

Reply:
> If the alpha setting is provided and set to false, then the canvas is forced to always be opaque.

It seems that `alpha: false` must be provided explicitly for the forced opaque behavior.